File: express/src/tests/api/endpoints/hello.test.ts
import supertest from "supertest";
import app from "../../../api";
import { crudUser } from "../../../models/crud";
import { connectDbs, closeDbs } from "../../../lib/clients";

const request = supertest(app);

beforeAll(async () => {
    await connectDbs();
});

afterAll(async () => {
    await closeDbs();
});

describe("GET /api/hello-world/", () => {
    it("responds with json", async () => {
        const response = await request
            .get("/api/hello-world/")
            .expect("Content-Type", /json/)
            .expect(200);

        expect(response.body).toEqual({ message: "Hello World!" });
    });
});

describe("GET /api/hello-world/user", () => {
    it("responds with json", async () => {
        const token = await crudUser.getBearer("mslimbeji@gmail.com");
        const response = await request
            .get("/api/hello-world/user")
            .set("Authorization", token)
            .expect("Content-Type", /json/)
            .expect(200);

        expect(response.body).toEqual({ message: "Hello Slim Beji!" });
    });
});

describe("GET /api/hello-world/admin", () => {
    it("responds with json", async () => {
        const token = await crudUser.getBearer("mslimbeji@gmail.com");
        const response = await request
            .get("/api/hello-world/admin")
            .set("Authorization", token)
            .expect("Content-Type", /json/)
            .expect(200);

        expect(response.body).toEqual({ message: "Hello Admin Slim Beji!" });
    });
});

File: backend-axum/src/tests/api/hello_test.rs
use crate::{
    models::cruds::CrudsUser,
    tests::api::utils::{get_content_type, parse_json, setup},
};
use axum::{
    body::Body,
    http::{Request, StatusCode},
};
use serde_json::json;
use tower::ServiceExt;

#[tokio::test]
async fn test_hello_world() {
    let (app, _) = setup().await;

    let response = app
        .oneshot(
            Request::builder()
                .method("GET")
                .uri("/api/hello-world")
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json = parse_json(response).await;
    assert_eq!(json, json!({"message": "Hello World!"}));
}

#[tokio::test]
async fn test_hello_user() {
    let (app, state) = setup().await;
    let cruds = CrudsUser::new(state);
    let token = cruds
        .get_bearer("mslimbeji@gmail.com")
        .await
        .expect("test error");

    let response = app
        .oneshot(
            Request::builder()
                .method("GET")
                .uri("/api/hello-world/user")
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json = parse_json(response).await;
    assert_eq!(json, json!({"message": "Hello Slim Beji!"}));
}

#[tokio::test]
async fn test_hello_admin() {
    let (app, state) = setup().await;
    let cruds = CrudsUser::new(state);
    let token = cruds
        .get_bearer("mslimbeji@gmail.com")
        .await
        .expect("test error");

    let response = app
        .oneshot(
            Request::builder()
                .method("GET")
                .uri("/api/hello-world/admin")
                .header("Authorization", token)
                .body(Body::empty())
                .expect("test error"),
        )
        .await
        .expect("test error");

    assert_eq!(response.status(), StatusCode::OK);
    assert!(get_content_type(&response).contains("application/json"));

    let json = parse_json(response).await;
    assert_eq!(json, json!({"message": "Hello Admin Slim Beji!"}));
}
